db/migrate/20230227162445_add_user_ref_to_coach.rb

```diff
@@ -0,0 +1,5 @@
+class AddUserRefToCoach < ActiveRecord::Migration[7.0]
+  def change
+    add_reference :coaches, :user, null: false, foreign_key: true
+  end
+end
```

Comment:
- [ ] This migration is not working:

![image](https://user-images.githubusercontent.com/57421823/224106169-534f59d0-77ac-4864-99f5-41984b0114f2.png)

This is causing the db to be unusable to new users.

## Explanation

In the second migration `20230227162445_add_user_ref_to_coach`, you are adding `user` references to the `coaches` table:

![image](https://user-images.githubusercontent.com/57421823/224107374-792e30a9-1e46-44a2-a829-e6be5410b900.png)

However, at the time of this migration, the `users` table does not exist

In fact, after further investigation, I can see that the 'users' table is not being created at all in any of the migrations (was it deleted?):

![image](https://user-images.githubusercontent.com/57421823/224107893-c8691790-6dd7-4165-b184-2b4d3eb7dec4.png)

The project may work for you because you already have the database set up, but not for new users.

## Solution and tips

# Solution 1 (recommended)

Drop the database first:

```
rails db:drop
```



You are now a new user.

Now, create the database again:

```
rails db:create
```

Remove all the migrations from the project (you can backup them elsewhere for reference).


At this point, the database is ready to receive migrations from the beginning.

Now, please generate and run your migrations one by one, using the ones that you backed up as a reference.  

The migrations must be new, the number in the file name is auto-generated and it's important (don't copy the file names, it won't work).

Remember, this whole issue was caused because the migration to add the users table does not exist.  Be sure to add that migration as well (according to your migrations order, it should be the second).

# Solution 2

Rollback (unrun) all your migrations up to 20230227154702 which is the only one that we can keep:

```bash
rails db:migrate VERSION=20230227154702 
```

Delete all other migrations in your project keeping only the first one.  You can make a copy of them for reference.

Now, re-generate and re-run all the other migrations one by one, following the advise above.

## Conclusion and further advise

Never delete a migration on a working database.  The migrations are set up as a history and if you delete one migration you are deleting one part of that history.  Rails tries to run the migrations one after the other, in order according to the number in the file name.  This means every new migration depends on the previous ones to be present.

If you notice that you messed up in one migration and you want to delete it or correct it:

- If this is the last migration, and it hasn't been run to the database, go ahead and delete it.
- If this is the last migration, and it was run to the database,  you can rollback (e.g. unrun) it first: `rake db:rollback STEP=1` then you can safely delete it.  You can also make a new migration that correct the last one instead of rolling back and deleting.
- If this is NOT the last migration, and the problem is NOT in the beginning of the history, you can rollback multiple migrations `rake db:rollback STEP=1`.  Substitute the "1" for the number of steps (migrations) you want to rollback.
- If this is NOT the last migration, and the problem is in the beginning of the history, you drop the database, delete all migrations, generate, and run them again (as in this case).

I'll do a little hacking in order to be able to continue with the review (load from schema).

_This happened about 2 weeks ago at work, and it was a mess to fix because we have hundreds of migrations.  You are lucky this happened to you here with just 8 migrations, and now know the solution._

Now that you know the solution, kindly fix your migrations. 💪 

I hope this is helpful and makes sense to you.